src/wallet/authhelper.cpp

```diff
@@ -0,0 +1,101 @@
+#include <random>
+#include <string>
+#include <map>
+
+#include "base58.h"
+#include "authhelper.h"
+
+struct AuthorizationHelper::Impl
+{
+    using rnd_type = size_t;
+    using code_type = std::string;
+    Impl()
+    : seed()
+    , generator(seed())
+    , rnd_dist(std::numeric_limits<rnd_type>::min(), std::numeric_limits<rnd_type>::max())
+    {}
+
+    bool authorize(code_type const & functionName, code_type code)
+    {
+        auto const fn_key_iter = key_store.find(functionName);
+        if(fn_key_iter == key_store.end())
+        {
+            generateAuthorizationCode(functionName);
+            return false;
+        }
+
+        if(code == toCode(fn_key_iter->second))
+        {
+            generateAuthorizationCode(functionName);
+            return true;
+        }
+
+        return false;
+    }
+
+    code_type generateAuthorizationCode(code_type const & functionName)
+    {
+        rnd_type key = rnd_dist(generator);
+        key_store[functionName] = key;
+        return toCode(key);
+    }
+
+private:
+    std::random_device seed;
+    std::mt19937 generator;
```

Comment:
There is (was?) a bug in mingw implementation of random_device
https://sourceforge.net/p/mingw-w64/bugs/338/
I need to double check if on Windows it returns always the same string

Reply:
Wow! Checking also.

Reply:
Yes, it's always the same
There are two options
1. Use boost::random_device. This requires adding new dependency (libboost-random-dev)
2. Switch to hashing wallet data (keys)

Reply:
Changed it to be according to the 1st option.
LOCK2(cs_main, pwalletMain->cs_wallet);
generator.seed(pwalletMain->GetHDChain().masterKeyID.GetUint64(0) ^ time(NULL));

Reply:
LGTM but please manually test if it gives different results on different Windows installations (even if the `time(NULL)` is exactly the same) before merging to master